Upgrade to sbt 1

Duhemm opened the pull request:
Projects from the community build to port to sbt1:

- [x] [algebra](https://github.com/dotty-staging/algebra/tree/sbt1)
- [x] [better-files](https://github.com/dotty-staging/better-files/tree/sbt1)
- [x] [scalacheck](https://github.com/dotty-staging/scalacheck/tree/sbt1)
- [x] [ScalaPB](https://github.com/dotty-staging/ScalaPB/tree/sbt1)
- [x] [scalatest](https://github.com/dotty-staging/scalatest/tree/sbt1)
- [x] [scopt](https://github.com/dotty-staging/scopt/tree/sbt1)
- [x] [squants](https://github.com/dotty-staging/squants/tree/sbt1)
- [x] [collections-strawman](https://github.com/dotty-staging/collection-strawman/tree/sbt1)

smarter commented:
Awesome! :)

smarter commented:
@ttreyer Could you please sign the Scala CLA at https://www.lightbend.com/contribute/cla/scala ? That way we'll finally be able to merge your code :smile:

smarter commented on project/Build.scala:
```diff
@@ -150,8 +151,8 @@ object Build {
     scalaCompilerBridgeSource :=
-      (dottyOrganization %% "dotty-sbt-bridge" % dottyVersion % Configurations.Component.name)
-      .artifacts(Artifact.sources("dotty-sbt-bridge").copy(url =
+      (dottyOrganization %% "dotty-sbt-bridge" % "NOT_PUBLISHED" % Configurations.Component.name)
```
A change was accidentally reverted here: we stopped using NOT_PUBLISHED in https://github.com/lampepfl/dotty/commit/78ac17ace84382322aadfa18334ba2a6b214d7ad

smarter commented:
Until we're ready to merge this PR, we should enable running all the sbt scripted tests on every push to make sure everything passes.

smarter commented:
https://github.com/sbt/zinc/pull/444 should be ported here too.

smarter commented on sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala:
```diff
@@ -49,52 +51,43 @@ object DottyPlugin extends AutoPlugin {
-            moduleID
-        }
-    }
+     implicit class DottyCompatModuleID(moduleID: ModuleID) {
```
extra space in indentation

smarter commented on sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala:
```diff
@@ -49,52 +51,43 @@ object DottyPlugin extends AutoPlugin {
   override def trigger = allRequirements
 
-  // Adapted from CrossVersionUtil#sbtApiVersion
+  // Adapted from CrossVersioconstant nUtil#sbtApiVersion
```
typos

smarter commented on project/Build.scala:
```diff
@@ -194,35 +197,39 @@ object Build {
-        ).map(_.withDottyCompat())
+          dottyOrganization % "dotty-library_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name,
+          dottyOrganization % "dotty-compiler_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name
+        )//.map(_.withDottyCompat(scalaVersion.value))
```
This should be reverted now that `withDottyCompat` has been ported.

Duhemm commented:
I addressed the comments so far. I've ported sbt/zinc#444 here.

All the scripted tests are already being run on every push for this PR.

I've also finished upgrading all the projects of the community build to sbt 1.0.3 and proposed to upstream the changes.

allanrenucci commented on project/Build.scala:
```diff
@@ -853,21 +853,22 @@ object Build {
-     // By default scripted tests use $HOME/.ivy2 for the ivy cache. We need to override this value for the CI.
-      ScriptedPlugin.scriptedLaunchOpts ++= ivyPaths.value.ivyHome.map("-Dsbt.ivy.home=" + _.getAbsolutePath).toList,
-      ScriptedPlugin.scripted := ScriptedPlugin.scripted.dependsOn(Def.task {
+      version := "0.1.6-SNAPSHOT",
```
1.8

smarter commented on project/Build.scala:
```diff
@@ -199,7 +199,7 @@ object Build {
         Seq(
           dottyOrganization % "dotty-library_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name,
           dottyOrganization % "dotty-compiler_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name
```
This line and the line above also need to be reverted to not use "_2.11"

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala:
```diff
@@ -254,8 +255,17 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
+          underlying :: ancestorTypes0
+        } else
+          ancestorTypes0
+      } catch {
```
The try/catch should only be needed around the call to ` linearizedAncestorTypes(cinfo)`.

Duhemm commented:
test performance please

allanrenucci commented:
One more thing to do is to create a new docker image from this PR, update the image in `drone.yml` and make sure this PR does not introduce a big slow down on the tests.

smarter commented on project/Build.scala:
```diff
@@ -862,7 +862,7 @@ object Build {
         baseDirectory.value / "../language-server/src/dotty/tools/languageserver/config",
       sbtPlugin := true,
-      version := "0.1.6-SNAPSHOT",
+      version := "0.1.8",
```
This is an incompatible change, I would bump to 0.2.0

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -90,13 +99,59 @@ class ExtractDependencies extends Phase {
+      def apply(x: Option[Symbol], t: Tree)(implicit ctx: Context) =
+        if (x.isDefined) x
+        else t match {
+          case moduleDef: Thicket =>
```
I don't think this case is ever going to be hit, objects in Dotty after Typer are represented by a TypeDef and a ValDef, the TypeDef case below will be used.

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -90,13 +99,59 @@ class ExtractDependencies extends Phase {
+    else {
+      val file = Option(sym.associatedFile)
+
+      Option(sym.associatedFile).flatMap {
```
I don't understand what this is for, if the symbol comes from source then `sym.associatedFile` will never be `NoSourceFile`.

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -105,34 +160,41 @@ class ExtractDependencies extends Phase {
-        }
-      } else if (depFile.file != currentSourceFile) {
-        ctx.sbtCallback.classDependency(extractedName(dep.enclosingClass), extractedName(currentClass), context)
+    val onSource = dep.sourceFile
```
Why was this changed? This is also very likely to do the wrong thing. If `dep` was loaded from TASTY, `dep.sourceFile` will return the source filename stored in the SoureFile attribute of the TASTY section. But here we're really interested in recording the dependency on the classfile where `dep` comes from, because if this classfile changes we'll have to recompile the current unit.

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -161,11 +223,13 @@ private final class NameUsedInClass {
-          builder.append(" in [")
-          otherScopes.forEach(new java.util.function.Consumer[UseScope]() {
-            def accept(scope: UseScope): Unit =
+          // Pickling tests fail when this is turned in an anonymous class
```
Could you open an issue so we don't forget about this?

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -182,13 +246,13 @@ private final class NameUsedInClass {
  *  inheritance" in the "Name hashing algorithm" section.
  */
-private class ExtractDependenciesCollector(implicit val ctx: Context) extends tpd.TreeTraverser { thisTreeTraverser =>
+private class ExtractDependenciesCollector(responsibleForImports: Symbol)(implicit val ctx: Context) extends tpd.TreeTraverser { thisTreeTraverser =>
```
Please add some documentation for this parameter

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -199,15 +263,18 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
    *  because it refers to these classes or something defined in them.
    *  This is always a superset of `topLevelInheritanceDependencies` by definition.
    */
```
Documentation needs to be updated since this is now a set of pair, same for the def just below.

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -199,15 +263,18 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
   private def addUsedName(enclosingSym: Symbol, name: Name) = {
-    val enclosingName = extractedName(enclosingSym)
+    val enclosingName = enclosingSym match {
+      case sym if sym == defn.RootClass => ExtractDependencies.extractedName(responsibleForImports)
```
Why use a pattern match and not just an if/else here?

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -276,19 +349,20 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
           case Thicket(Ident(name) :: Ident(rename) :: Nil) =>
             addImported(name)
-            if (rename ne nme.WILDCARD)
+            if (rename ne nme.WILDCARD) {
```
Unnecessary brackets

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -241,10 +314,10 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
+    _topLevelInheritanceDependencies += ((currentClass, sym.topLevelClass))
 
-  private object PatMatDependencyTraverser extends ExtractTypesCollector {
+  private class PatMatDependencyTraverser(ctx0: Context) extends ExtractTypesCollector(ctx0) {
```
Why not `implicit ctx: Context` ? And at the same time, the `implicit ctx: Context` argument to method in this class should be removed.

smarter commented on compiler/src/dotty/tools/dotc/sbt/ThunkHolder.scala:
```diff
@@ -25,97 +25,8 @@ private[sbt] trait ThunkHolder {
-    SafeLazyWrapper.strict(t)
-}
-
-/** Wrapper around SafeLazy implementations.
```
So great that we can get rid of this stuff :smile: :fire:

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala:
```diff
@@ -182,32 +209,70 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
+      name, acc, modifiers, anns, defType, api.SafeLazy.strict(selfType), api.SafeLazy.strict(structure), Constants.emptyStringArray,
+      childrenOfSealedClass, topLevel, tparams)
+
+    // if (name.toString.contains("DottyPredef")) {
```
Remove debug code

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala:
```diff
@@ -182,32 +209,70 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
+
+    allNonLocalClassesInSrc += cl
+
+    val javaPlatform = ctx.platform.asInstanceOf[JavaPlatform]
```
Please avoid the cast here, we don't want the compiler to crash when run with the Scala.js backend. I think we should just move `hasJavaMainMethod` outside of `JavaPlatform`

sjrd replied:
Yes, especially since Scala.js supports exactly the same main methods as Scala/JVM!

smarter commented on compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -86,38 +160,84 @@ class ExtractDependencies extends Phase {
+    // }
+}
+
+private final class NameUsedInClass {
```
All the added class/object in this PR should get a documentation comment.

jvican commented:
Is there any chance this can be merged and released together with 0.5.0-RC1? I'd like to try cross-compiling bloop to Dotty. For now, I'm doing it locally, but ideally I'd have an official version to depend on so that the CI passes my changes.

smarter commented:
> Is there any chance this can be merged and released together with 0.5.0-RC1?

No, that would require delaying the release too much. But once this PR is merged you'll be able to use our nightly builds in your project, using a nightly build of Dotty is just as easy as using a release.

allanrenucci commented:
We had a regression with `sbt-pack` #3124: the generated jar contained the non-bootstrapped version. Since you are updating the version, it would be good to know if the issue is still present.
/cc @liufengyun

allanrenucci commented on .drone.yml:
```diff
@@ -41,15 +41,25 @@ pipeline:
+    image: lampepfl/dotty:2017-10-20
+    commands:
+      - cp -R . /tmp/5/ && cd /tmp/5/
+      - ./project/scripts/sbt "sbt-dotty/scripted source-dependencies/*2of2"
```
It would be great if we did not introduce this huge regression (3x) in the time needed to run the scripted tests.
@Duhemm says it is because the bridge is recompiled for each scripted test. @smarter do you have an idea how this could be fixed?

smarter replied:
Complain to the sbt people: https://github.com/sbt/sbt/issues/3469 Maybe ask someone who has a lightbend subscription? :)

smarter replied:
Alternatively, we could try copy-pasting the custom implementation of scripted used by zinc, it even has parallelism and fancy colors contributed by @jvican :P. https://github.com/sbt/zinc/blob/1.x/project/Scripted.scala https://github.com/sbt/zinc/tree/1.x/internal/zinc-scripted/src/test/scala/sbt/internal/inc

liufengyun commented:
Good catch @allanrenucci : `0.9.1` is the version where we had the regression. The latest version for `sbt-pack` is `1.0.4`. We can run `dist-bootstrapped/pack`, if the generated jars are good, then there's no worry.

allanrenucci commented:
Unfortunately `dist-bootstrapped/pack` generates the non bootstrapped jars

xerial commented:
@allanrenucci @liufengyun Released sbt-pack 0.10.0, which fixes an issue of including too many modules. https://github.com/xerial/sbt-pack/issues/67

xerial commented:
Released sbt-pack 0.10.1, which fixes a bug if you have multiple nested projects in build.sbt. Please use this version. Thanks.

allanrenucci commented:
Thanks @xerial! I can confirm this fixes our issue

OlivierBlanvillain commented:
I'm closing this PR given that it's been rebased in #3687. @allanrenucci next time just merge master into the existing PR, rebasing doesn't scale to multiple committers and long standing PRs...
